### PR TITLE
#895 Fix truncated last frame of SpriteAnimation

### LINKED ISSUE #895: AnimationGroup final frame timing off
# Current bug behaviour
Second animation in AnimationGroup has an incorrect duration for the final frame
[Here's a link to a video demoing the bug](https://youtu.be/iR6BHEkeat4)

# Expected behaviour
Same duration for all frames

# Steps to reproduce

```
import 'package:flame/components.dart';
import 'package:flame/game.dart';
import 'package:flame/gestures.dart';
import 'package:flame/sprite.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(App());
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Animations',
      theme: ThemeData(
        brightness: Brightness.light,
        /* light theme settings */
        primarySwatch: Colors.lime,
      ),
      darkTheme: ThemeData(
        brightness: Brightness.dark,
        /* dark theme settings */
        primarySwatch: Colors.lime,
      ),
      themeMode: ThemeMode.dark,
      home: TestAnimationPage(),
    );
  }
}

class TestAnimationPage extends StatefulWidget {
  @override
  _TestAnimationPageState createState() => _TestAnimationPageState();
}

class _TestAnimationPageState extends State<TestAnimationPage> {
  @override
  Widget build(BuildContext context) {
    Size screenSize = MediaQuery.of(context).size;
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text('Animation Group Test'),
      ),
      body: Stack(
        alignment: Alignment.topCenter,
        children: [
          GameWidget(game: TestAnimationGame(screenSize)),
          _overlayWidget(),
        ],
      ),
    );
  }
  Widget _overlayWidget() {
    return const Padding(
      padding: const EdgeInsets.all(8.0),
      child: Text(
        'Tap on Bob',
        style: TextStyle(color: Colors.white),
      ),
    );
  }
}

enum BobAnimationState {
  idle,
  jump,
}

class Bob extends PositionComponent with Tappable {
  late SpriteAnimationGroupComponent animationGroup;
  Future<SpriteAnimation> Function(String path, SpriteAnimationData data) loadSpriteAnimation;

  Bob({required this.loadSpriteAnimation}) {
    size = Vector2(50, 50);
    this.debugColor = Colors.green;
  }

  @override
  Future<void>? onLoad() async {
    const String sheetFile = 'numbers.png';
    double stepTime = 1;
    SpriteAnimation idle = await loadSpriteAnimation(
      sheetFile,
      SpriteAnimationData.sequenced(
        amount: 4,
        textureSize: Vector2(8, 8),
        texturePosition: Vector2(0, 0),
        stepTime: stepTime,
      ),
    );
    SpriteAnimation jump = await loadSpriteAnimation(
      sheetFile,
      SpriteAnimationData.sequenced(
        amount: 4,
        textureSize: Vector2(8, 8),
        texturePosition: Vector2(0, 8),
        stepTime: stepTime,
        loop: false,
      ),
    );
    animationGroup = SpriteAnimationGroupComponent(
      current: BobAnimationState.idle,
      animations: {
        BobAnimationState.idle: idle,
        BobAnimationState.jump: jump,
      },
      size: size,
      position: position,
    );
    jump.onComplete = () {
      animationGroup.current = BobAnimationState.idle;
      jump.reset();
    };
    return super.onLoad();
  }

  @override
  void update(double dt) {
    animationGroup.update(dt);
    super.update(dt);
  }

  void render(Canvas canvas) {
    animationGroup.render(canvas);
    super.render(canvas);
  }

  @override
  bool onTapDown(TapDownInfo info) {
    animationGroup.current = BobAnimationState.jump;
    animationGroup.animations[BobAnimationState.idle]!.reset();
    return super.onTapDown(info);
  }
}

class TestAnimationGame extends BaseGame with HasTappableComponents {
  late Bob bob;
  final Size screenSize;

  TestAnimationGame(this.screenSize);

  @override
  Future<void> onLoad() async {
    bob = Bob(loadSpriteAnimation: loadSpriteAnimation);
    bob.position = Vector2(screenSize.width / 2 - bob.width / 2, 100);
    add(bob);
    return super.onLoad();
  }

  @override
  void render(Canvas canvas) {
    super.render(canvas);
  }

  @override
  void update(double dt) {
    super.update(dt);
  }

  @override
  Color backgroundColor() => Colors.black;
}

```

# Flutter doctor output
```
[√] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.19042.1110], locale en-US)
    • Flutter version 2.2.3 at C:\flutter
    • Framework revision f4abaa0735 (4 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2) 
    • Android SDK at C:\Users\Nate\AppData\Local\Android\sdk
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = C:\Users\Nate\AppData\Local\Android\sdk
    • Java binary at: E:\Programs\AndroidStudio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe      

[√] Android Studio (version 3.6)
    • Android Studio at E:\Programs\AndroidStudio
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[√] VS Code (version 1.58.2)
    • VS Code at C:\Users\Nate\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[√] Connected device (3 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 92.0.4515.107
    • Edge (web)                         • edge          • web-javascript • Microsoft Edge 92.0.902.55

• No issues found!
```

# More environment information
* Flame version: ^1.0.0-releasecandidate.13
* Platform affected: Reproduced on Android and Web, didn't try any other platforms

# Here's a spritesheet to test with
numbers

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

The code in `SpriteAnimation` used to exit early when it found itself on the last frame with `loop=false`. The new behavior will wait until the last animation frame is ready to roll over to the next frame before deciding to officially stop the animation and call the `onComplete` callback.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage.
- [ ] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [ ] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems. <-- I don't think these scripts exist anymore.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

Closes #895

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

**Review by spydon — APPROVED** [review]
LGTM! Thanks for your contribution. :)

**Comment by spydon** [comment]
Can you rebase on `main` @st-pasha?

File: packages/flame/test/components/sprite_animation_component_test.dart [hunk 1]
```diff
@@ -138,5 +138,41 @@ void main() async {
       game.update(0.1);
       expect(game.components.length, 1);
     });
+
+    test('Last animation frame is not skipped', () {
+      // See issue #895
```
Comment (erickzanardo) [inline packages/flame/test/components/sprite_animation_component_test.dart:143]: We could have the link for the issue here instead of only the number

**Review by erickzanardo — APPROVED** [review]
One small bit, but LGTM already. Thanks for the fix

File: packages/flame/test/components/sprite_animation_component_test.dart (hunk 1 above)
Comment (st-pasha) [reply packages/flame/test/components/sprite_animation_component_test.dart]: done

**Pull request merged**